storm-client/src/jvm/org/apache/storm/grouping/LoadAwareShuffleGrouping.java

```diff
@@ -52,25 +78,92 @@ public void prepare(WorkerTopologyContext context, GlobalStreamId stream, List<I
     }
 
     @Override
+    public void refreshLoad(LoadMapping loadMapping) {
+        updateRing(loadMapping);
+    }
+
+    @Override
     public List<Integer> chooseTasks(int taskId, List<Object> values, LoadMapping load) {
-        if ((lastUpdate + 1000) < System.currentTimeMillis()) {
-            int local_total = 0;
-            for (int i = 0; i < targets.length; i++) {
-                int val = (int)(101 - (load.get(targets[i]) * 100));
-                loads[i] = val;
-                local_total += val;
+        int rightNow;
+        while (true) {
+            rightNow = current.incrementAndGet();
+            if (rightNow < CAPACITY) {
+                return rets[choices[rightNow]];
+            } else if (rightNow == CAPACITY) {
+                current.set(0);
```

Comment:
I borrowed this from ShuffleGrouping, and even I'm not clear whether it makes race condition, I think it is acceptable since race condition scenarios don't incur out of index, just letting some threads selecting same index and maybe skip some indices.
Moreover this patch contains tests which addresses multi-thread safety.

We can still replace `set` with `getAndSet` to make it fully thread-safe, but we need to do more experiments if we would want to. Same applies to ShuffleGrouping.